File: Java/2268/src/Main.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {
    static long[] tree;

    public static void main(String[] args) throws IOException, NumberFormatException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        tree = new long[n * 4];

        StringBuffer sb = new StringBuffer();
        for(int i = 0; i < m; i ++){
            st = new StringTokenizer(br.readLine());

            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());

            if (a == 0){
                if(b > c){
                    int temp = b;
                    b = c;
                    c = temp;
                }
                sb.append(query(1, n, 1, b,c)).append("\n");
            } else if (a == 1){
                update(1, n,1, b, c);
            }
        }
        bw.write(sb.toString());
        bw.flush();
        bw.close();
        br.close();
    }
    public static long query(int start, int end, int node, int left, int right){
        if (left > end || start > right) return 0;

        if (left <= start && end <= right) return tree[node];

        int mid = (start + end) / 2;
        long left_ = query(start, mid, node * 2, left, right);
        long right_ = query(mid + 1, end, node * 2 + 1, left, right);
        return left_ + right_;
    }

    public static long update(int start, int end, int node, int idx, int val){
        if (idx < start || idx > end) return tree[node];

        if (start == end) {
            tree[node] = val;
            return tree[node];
        }

        int mid = (start + end) / 2;
        tree[node] = update(start, mid, node * 2, idx, val) + update(mid + 1, end, node * 2 + 1, idx, val);
        return tree[node];
    }
}

File: python/2268/main.py
import sys
input = sys.stdin.readline


class Node:
    def __init__(self, start, end):
        self.start, self.end = start, end
        self._left, self._right = None, None
        self.data = 0

    @property
    def mid(self):
        return (self.start + self.end) // 2

    @property
    def left(self):
        self._left = self._left or Node(self.start, self.mid)
        return self._left

    @property
    def right(self):
        self._right = self._right or Node(self.mid +  1, self.end)
        return self._right

    def query(self, left, right):
        if left <= self.start and self.end <= right:
            return self.data
        if self.end < left or right < self.start:
            return 0
        left_ = self.left.query(left, right)
        right_ = self.right.query(left, right)
        return left_ + right_

    def update(self, idx, diff):
        if idx < self.start or self.end < idx:
            return
        self.data += diff
        if not self.start == self.end:
            self.left.update(idx, diff)
            self.right.update(idx, diff)


n, m = map(int, input().split())
tree = Node(1, n)
data = [0] * (n + 1)
for i in range(m):
    order, *args = map(int, input().split())
    if order == 0:
        print(tree.query(args[0], args[1]))
    elif order == 1 and data[args[0]] != args[1]:
        diff = args[1] - data[args[0]]
        tree.update(idx=args[0], diff=diff)
        data[args[0]] = args[1]